lib/src/main/java/space/taran/arklib/domain/index/RootIndex.kt

```diff
@@ -0,0 +1,137 @@
+import java.nio.file.Path
+import kotlin.io.path.ExperimentalPathApi
+
+class ResourceUpdates (
+    val deleted: Map<ResourceId, LostResource>,
+    val added: Map<ResourceId, NewResource>
+)
+
+data class LostResource(val path: Path, val resource: Resource)
+
+data class NewResource(val path: Path, val resource: Resource)
+
+/**
+ * [RootIndex] is a type of index backed by storage file.
+ * It is made to store ids of all resources in a "root folder",
+ * which is:
+ * 1) a unit of file synchronization,
+ *    i.e. we either sync all of its content or none;
+ * 2) a mounting point in resource space,
+ *    i.e. total resource space is a union of resources
+ *    from all root folders.
+ * @param path A path to the root folder.
+ *
+ * See also [IndexAggregation] and [IndexProjection].
+ */
+@OptIn(ExperimentalPathApi::class)
+class RootIndex(val path: Path): ResourceIndex {
+
+    private val mutex = Mutex()
+    private val mutUpdates = MutableSharedFlow<ResourceUpdates>()
+
+    private val resourceById: MutableMap<ResourceId, Resource> = mutableMapOf()
+    private val pathById: MutableMap<ResourceId, Path> = mutableMapOf()
+
+    private fun wrap(update: RawUpdates): ResourceUpdates {
+        val deleted = update.deleted.associateWith { id ->
+            val path = pathById[id]!!
+            val resource = resourceById[id]!!
+            LostResource(path, resource)
+        }
+
+        val added = update.added
+            //we can't present empty resources
+            .mapNotNull { (id, path) ->
+                val resource: Resource = Resource.compute(id, path)
+                    .onFailure { error ->
+                        Log.e(
+                            LogTags.RESOURCES_INDEX,
+                            "Couldn't compute resource by path $path: $error"
+                        )
+
+                        return@mapNotNull null
+                    }.getOrThrow()
+
+                id to NewResource(path, resource)
+            }.toMap()
+
+        return ResourceUpdates(deleted, added)
+    }
+
+    init {
+        if (!BindingIndex.load(path)) {
+            Log.e(
+                RESOURCES_INDEX,
+                "Couldn't provide index from $path"
+            )
+            throw NotImplementedError()
+        }
+
+        //id2path should be used in order to filter-out duplicates
+        //path2id could contain several paths for the same id
+        BindingIndex.id2path(path)
+            .forEach { (id, path) ->
+                Resource.compute(id, path)
+                    .onFailure { error ->
+                        Log.e(
+                            RESOURCES_INDEX,
+                            "Couldn't compute resource by path $path: $error"
+                        )
+                    }
+                    .onSuccess { resource ->
+                        pathById[id] = path
+                        resourceById[id] = resource
+                    }
+            }
+    }
```

Comment:
I think this should be suspend method
Because it is heavy operation that is performed implicitly when constructor is called and it can block main thread

Reply:
Good catch